### ISSUE TITLE
Crash when using interfaces

### ISSUE THREAD

**Issue by olivernybroe** [nunomaduro/larastan#1997]
<!-- First, thank you for reporting a bug. That takes time and we appreciate that! -->

- Larastan Version: 2.9.8
- Laravel Version: 11.16.0

### Description

When defining a relationship on an interface and calling methods on that relationship, larastan crashes because it tries to get `newEloquentBuilder` from the interface. 

```
Internal error: Method newEloquentBuilder() was not found in reflection of class My\Package\Contracts\User. while analysing file /path/my-package/src/Actions/RemoveTeamMember.php   
```

<details>

<summary>Stacktrace</summary>

```

## phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(564)                                                                                                                                                                                                                                                                       
     #0 /path/my-package/vendor/larastan/larastan/src/Methods/BuilderHelper.php(220): PHPStan\Reflection\ClassReflection->getNativeMethod('newEloquentBuil...')                                                                                                                                                                                                                   
     #1 /path/my-package/vendor/larastan/larastan/src/Methods/RelationForwardsCallsExtension.php(79): Larastan\Larastan\Methods\BuilderHelper->determineBuilderName('My\\Packa...')                                                                                                                                                                                        
     #2 /path/my-package/vendor/larastan/larastan/src/Methods/RelationForwardsCallsExtension.php(39): Larastan\Larastan\Methods\RelationForwardsCallsExtension->findMethod(Object(PHPStan\Reflection\ClassReflection), 'whereKey')                                                                                                                                                
     #3 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(512): Larastan\Larastan\Methods\RelationForwardsCallsExtension->hasMethod(Object(PHPStan\Reflection\ClassReflection), 'whereKey')                                                                                                                                          
     #4 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php(558): PHPStan\Reflection\ClassReflection->hasMethod('whereKey')                                                                                                                                                                                                                       
     #5 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(3852): PHPStan\Type\ObjectType->hasMethod('whereKey')                                                                                                                                                                                                                          
     #6 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(3880): PHPStan\Analyser\MutatingScope->filterTypeWithMethod(Object(PHPStan\Type\Generic\GenericObjectType), 'whereKey')                                                                                                                                                        
     #7 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1602): PHPStan\Analyser\MutatingScope->methodCallReturnType(Object(PHPStan\Type\Generic\GenericObjectType), 'whereKey', Object(PhpParser\Node\Expr\MethodCall))                                                                                                                
     #8 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1608): PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}()                                                                                                                                                                                                            
     #9 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(599): PHPStan\Analyser\MutatingScope->resolveType('$team->users()-...', Object(PhpParser\Node\Expr\MethodCall))                                                                                                                                                                
     #10 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1602): PHPStan\Analyser\MutatingScope->getType(Object(PhpParser\Node\Expr\MethodCall))                                                                                                                                                                                        
     #11 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1608): PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}()                                                                                                                                                                                                           
     #12 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(599): PHPStan\Analyser\MutatingScope->resolveType('$team->users()-...', Object(PhpParser\Node\Expr\MethodCall))                                                                                                                                                               
     #13 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(825): PHPStan\Analyser\MutatingScope->getType(Object(PhpParser\Node\Expr\MethodCall))                                                                                                                                                                                         
     #14 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(599): PHPStan\Analyser\MutatingScope->resolveType('$deletes = $tea...', Object(PhpParser\Node\Expr\Assign))                                                                                                                                                                   
     #15 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(1740): PHPStan\Analyser\MutatingScope->getType(Object(PhpParser\Node\Expr\Assign))                                                                                                                                                                                        
     #16 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(696): PHPStan\Analyser\NodeScopeResolver->findEarlyTerminatingExpr(Object(PhpParser\Node\Expr\Assign), Object(PHPStan\Analyser\MutatingScope))                                                                                                                            
     #17 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(448): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Expression), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))                                                                     
     #18 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(624): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\ClassMethod), Array, Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))                                                            
     #19 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(448): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\ClassMethod), Object(PHPStan\Analyser\MutatingScope), Object(PHPStan\Node\ClassStatementsGatherer), Object(PHPStan\Analyser\StatementContext))                                       
     #20 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(761): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\Class_), Array, Object(PHPStan\Analyser\MutatingScope), Object(PHPStan\Node\ClassStatementsGatherer), Object(PHPStan\Analyser\StatementContext))                                    
     #21 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(448): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Class_), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))                                                                         
     #22 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(731): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\Namespace_), Array, Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))                                                             
     #23 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(416): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Namespace_), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))                                                                     
     #24 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(179): PHPStan\Analyser\NodeScopeResolver->processNodes(Array, Object(PHPStan\Analyser\MutatingScope), Object(Closure))                                                                                                                                                         
     #25 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(139): PHPStan\Analyser\FileAnalyser->analyseFile('/Users/oliverny...', Array, Object(PHPStan\Rules\LazyRegistry), Object(PHPStan\Collectors\Registry), NULL)                                                                                                                   
     #26 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111): PHPStan\Command\WorkerCommand::PHPStan\Command\{closure}(Array)                                                                                                                                                                                   
     #27 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/clue/ndjson-react/src/Decoder.php(117): _PHPStan_18cddd6e5\Evenement\EventEmitter->emit('data', Array)                                                                                                                                                                                                
     #28 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111): _PHPStan_18cddd6e5\Clue\React\NDJson\Decoder->handleData(Array)                                                                                                                                                                                   
     #29 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/Util.php(62): _PHPStan_18cddd6e5\Evenement\EventEmitter->emit('data', Array)                                                                                                                                                                                                         
     #30 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111): _PHPStan_18cddd6e5\React\Stream\Util::_PHPStan_18cddd6e5\React\Stream\{closure}('{"action":"anal...')                                                                                                                                             
     #31 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/DuplexResourceStream.php(168): _PHPStan_18cddd6e5\Evenement\EventEmitter->emit('data', Array)                                                                                                                                                                                        
     #32 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(201): _PHPStan_18cddd6e5\React\Stream\DuplexResourceStream->handleData(Resource id #6101)                                                                                                                                                                   
     #33 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(173): _PHPStan_18cddd6e5\React\EventLoop\StreamSelectLoop->waitForStreamActivity(NULL)                                                                                                                                                                      
     #34 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(99): _PHPStan_18cddd6e5\React\EventLoop\StreamSelectLoop->run()                                                                                                                                                                                                                
     #35 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Command/Command.php(259): PHPStan\Command\WorkerCommand->execute(Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Output\ConsoleOutput))                                                                     
     #36 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(870): _PHPStan_18cddd6e5\Symfony\Component\Console\Command\Command->run(Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Output\ConsoleOutput))                                              
     #37 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(261): _PHPStan_18cddd6e5\Symfony\Component\Console\Application->doRunCommand(Object(PHPStan\Command\WorkerCommand), Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Output\ConsoleOutput))  
     #38 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(157): _PHPStan_18cddd6e5\Symfony\Component\Console\Application->doRun(Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_18cddd6e5\Symfony\Component\Console\Output\ConsoleOutput))                                                
     #39 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(124): _PHPStan_18cddd6e5\Symfony\Component\Console\Application->run()                                                                                                                                                                                                                            
     #40 phar:///path/my-package/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(125): _PHPStan_18cddd6e5\{closure}()                                                                                                                                                                                                                                                             
     #41 /path/my-package/vendor/phpstan/phpstan/phpstan(8): require('phar:///Users/o...')                                                                                                                                                                                                                                                                                        
     #42 /path/my-package/vendor/bin/phpstan(119): include('/Users/oliverny...')        
```

</details>


### Laravel code where the issue was found

```php
$team->users()->whereKey($user->getKey())->delete();
```

Where user that is typehinted in the `users()` relationship is an interface

**Comment by calebdw** [comment]
What does the interface look like? You'll likely have to use an intersection type as the TRelatedModel is expected to be a Model

**Comment by olivernybroe** [comment]
Whoops, sorry forgot to include that. This is the interface

```php
/**
 * @template TEAM of Team
 * @template ORG of Organization
 */
interface User
{
    /**
     * @return BelongsToMany<TEAM>
     */
    public function teams(): BelongsToMany;

    /**
     * @return BelongsToMany<ORG>
     */
    public function organizations(): BelongsToMany;

    /**
     * Get the value of the model's primary key.
     *
     * @return mixed
     */
    public function getKey();
}

```

The Team and Organization classes referenced are also interfaces.

**Comment by olivernybroe** [comment]
And yes, just tested, doing an intersection type with `Model` fixes it which I will do to fix my issue. 

Guessing we are keeping this ticket open as this scenario actually results in a crash and not just wrong types?

**Comment by calebdw** [comment]
I don't think you need `getKey` on the interface as that's already on the model.

I can take a look and try to return a never type or something if the method is not found

**Comment by calebdw** [comment]
@olivernybroe, can you please show how you are defining the relationship?

**Comment by olivernybroe** [comment]
Sure, I can share you how my code looks now after I fixed the problem :) 

So my interface looks like this 
```php
/**
 * @template TEAM of Model&Team
 * @template ORG of Model&Organization
 *
 * @property OrganizationUser<ORG, $this>|null $organization_user
 * @property TeamUser<TEAM, $this>|null $team_user
 */
interface User
{
    /**
     * @return BelongsToMany<TEAM>
     */
    public function teams(): BelongsToMany;

    /**
     * @return BelongsToMany<ORG>
     */
    public function organizations(): BelongsToMany;

    /**
     * Get the value of the model's primary key.
     *
     * @return mixed
     */
    public function getKey();
}
```

and model
```php
/**
 * @template TEAM of Team
 * @template ORG of Organization
 *
 * @implements UserContract<TEAM, ORG>
 */
class User extends Model implements UserContract
{
    use HasFactory;

    /**
     * @return BelongsToMany<TEAM>
     */
    public function teams(): BelongsToMany
    {
        return $this->belongsToMany(
            OrganizationConfiguration::$teamModel,
            TeamUser::class,
            'user_id',
            'team_id'
        )->using(TeamUser::class)
            ->withPivot('role_id')
            ->as('team_user')
            ->withTimestamps();
    }

    /**
     * @return BelongsToMany<ORG>
     */
    public function organizations(): BelongsToMany
    {
        return $this->belongsToMany(
            OrganizationConfiguration::$organizationModel,
            OrganizationUser::class,
            'user_id',
        )->using(OrganizationUser::class)
            ->withPivot('role_id')
            ->as('organization_user')
            ->withTimestamps();
    }
}
```

These two classes lives in a package and then in my application, I have something like this
```
/**
 * @extends User<TeamImpl, OrganizationImpl>
 */
class UserImpl extends User
{

}
```

**Comment by canvural** [comment]
This doesn't make too much sense to me 🤔  And I'm surprised PHPStan does not throw more errors. Because the generic type for any relation only accepts child classes of `Model` (`T of Model` in short) So putting an interface there should already trigger an error.

Anyway, fixed the crash with 15348f1e820e80397788371d8e228ef0249f45fe 👍🏽
